### PR TITLE
Implements Config API

### PR DESCRIPTION
**Pull request by danehans**
- Adds `server` command to serve Envoy Gateway.
- Adds logr logging framework and uses zapr as the default logging implementation.
- Adds initial Kubernetes provider implementation.
- Adds make target to run EG locally against a k8s cluster.

Signed-off-by: danehans <daneyonhansen@gmail.com>

### CODE REVIEW

**Review by kflynn — COMMENTED** [review]
Per the community meeting today, maybe crank up the code coverage – but yeah, cool to see the framework in place!

File: hack/run-kube-local.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+set -euo pipefail
```
Comment (arkodg) [inline hack/run-kube-local.sh:1]: can this be converted into a target in https://github.com/envoyproxy/gateway/blob/main/tools/make/kube.mk ?

File: internal/cmd/server.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	return cmd
+}
+
+// serve Envoy Gateway using log as the logger.
+func serve(log logr.Logger) error {
+	cfg := &envoygateway.Config{Logger: log}
+
+	// Load the config file.
+	data, err := os.ReadFile(configPath)
```
Comment (arkodg) [inline internal/cmd/server.go:40]: with the current approach the process will exit early if a user only runs `envoy-gateway server` and `/etc/envoy-gateway/config.yaml` is missing . Instead should `config-path` default to "", and if empty serve can use system defaults ?

File: internal/cmd/server.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+				log.Error(err, "failed to cast config file", "name", configPath)
+				os.Exit(1)
+			}
+			// Set defaults for unset fields
+			eg.SetDefaults()
+			cfg.EnvoyGateway = eg
+		}
+	}
+
+	switch cfg.EnvoyGateway.Provider.Type {
```
Comment (arkodg) [inline internal/cmd/server.go:71]: nit: separate function for configuring/starting provider which creates the right template for other components

**Review by youngnick — APPROVED** [review]
LGTM, nice work.

File: hack/actions/install-kubernetes-toolchain.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
+#! /usr/bin/env bash
```
Comment (arkodg) [inline hack/actions/install-kubernetes-toolchain.sh:1]: can these be incorporated into https://github.com/envoyproxy/gateway/blob/main/tools/make/kube.mk which is doing something similar (installing kube tools)

File: .github/workflows/build_and_test.yaml [hunk 5]
```diff
@@ -38,6 +38,11 @@ jobs:
         uses: actions/setup-go@v3
         with:
           go-version: ${{ env.GO_VERSION }}
+      - name: deps
```
[the remark this answers is not in the document]
Comment (danehans) [reply .github/workflows/build_and_test.yaml]: @arkodg commit `de93191` removes this target. Instead, the existing `go.test.coverage` target calls the `envtest` target to install the [setup-envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/tools/setup-envtest) bin and sets the `KUBEBUILDER_ASSETS` env var used by the kube provider integration tests.

[the remark this answers is not in the document]
Comment (arkodg) [reply .github/workflows/build_and_test.yaml]: ah fine for now, but we should consider adding a separate make target for kube integration tests in the future

**Comment by danehans** [comment]
Merging since @youngnick already /approve.

**Pull request merged**